apps/designer/app/designer/shared/assets/use-assets.tsx

```diff
@@ -0,0 +1,316 @@
+
+export const AssetsProvider = ({ children }: { children: ReactNode }) => {
+  const [project] = useProject();
+  const [stateAssets, setAssets] = useAssetsState();
+  const { load, data: serverAssets } = useFetcher<Asset[]>();
+  const submit = usePersistentFetcher();
+  const assetsRef = useRef(stateAssets);
+
+  const action = project && restAssetsPath({ projectId: project.id });
+  assetsRef.current = stateAssets;
+
+  useEffect(() => {
+    if (action && stateAssets.length === 0) {
+      load(action);
+    }
+  }, [action, stateAssets.length, load]);
+
+  useEffect(() => {
+    if (serverAssets !== undefined) {
+      const nextAssets = updateStateAssets(stateAssets, serverAssets);
+
+      if (nextAssets !== stateAssets) {
+        setAssets(nextAssets);
+      }
+    }
+  }, [serverAssets, stateAssets, setAssets]);
+
+  const handleDeleteAfterSubmit = (data: UploadData) => {
+    // We need to remove assets only at updateStateAssets.
+    // We can't remove it here optimistically because the previous load
+    // can return it as "updated" because of the async operation nature
+    if (action) {
+      load(action);
+    }
+
+    if (data.status === "error") {
+      // We don't know what's wrong, remove the "deleting" status from assets and wait for the load to fix it
+      const assets = assetsRef.current;
+      const nextAssets = [...assets].map((asset) => {
+        if (asset.status === "deleting") {
+          const newAsset = { ...asset, status: "uploaded" } as Asset;
+          return newAsset;
+        }
+
+        return asset;
+      });
+
+      setAssets(nextAssets);
+
+      return toastUnknownFieldErrors(normalizeErrors(data.errors), []);
+    }
+  };
+
+  const handleAfterSubmit = (previewAssetId: string) => (data: UploadData) => {
+    const assets = assetsRef.current;
+
+    if (action) {
+      load(action);
+    }
+
+    if (data.status === "error") {
+      // We don't know what's wrong, remove preview asset and wait for the load to fix it
+      setAssets(assets.filter((asset) => asset.id !== previewAssetId));
+
+      return toastUnknownFieldErrors(normalizeErrors(data.errors), []);
+    }
+
+    warnOnce(
+      data.uploadedAssets?.length !== 1,
+      "Expected exactly 1 uploaded asset"
+    );
+
+    const uploadedAsset = data.uploadedAssets?.[0];
+
+    if (uploadedAsset === undefined) {
+      warnOnce(true, "An uploaded asset is undefined");
+      toast.error("Could not upload an asset");
+
+      // @todo better to create ErrorAsset type and show asset with error
+      setAssets(assets.filter((asset) => asset.id !== previewAssetId));
+      return;
+    }
+
+    // Optimistically update preview asset with the uploaded asset
+    const nextAssets = [...assets];
+
+    const index = nextAssets.findIndex(
+      (nextAsset) => nextAsset.id === previewAssetId
+    );
+
+    if (index !== -1) {
+      nextAssets[index] = { ...uploadedAsset, status: "uploading" };
+    }
```

Comment:
What are we updating? Isn't `status=uploading` at this point? And we are setting it to `uploading` again? Just updating data, but is it necessary?

Reply:
Looking on this I found the other bug, at this point load can already deliver asset with real id, and we are getting duplicate here if it is. (this code was to sync previewId and id) after discussion we decided to generate uuid on the client, otherwise race conditions like this is hard to test/find and detect.

Reply:
Now id is generated on a client, should be no race effect, comment has changed